### PR TITLE
Bug #5030 Generation of CFD data not working

### PR DESCRIPTION
**Pull request by jovuribe**
#### Description

Addressing issue #5030 with random generation of CFD data not working

In RandomDataGeneratorProgram.cs, var delistDate will produce a minValue greater than the maxValue at settings.Start.AddMonths(6) if the time difference between settings.Start and settings.End is less than 6 months. I addressed this issue by replacing this operation with a Datetime object that is halfway between settings.Start and settings.End.

#### Related Issue

[https://github.com/QuantConnect/Lean/issues/5030](url)

#### Motivation and Context

I spent some time working with Toolbox during the summer and this issue was brought to my attention. It solves an issue with random data generation for CFD data.

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

I tested this on my private fork.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [] I have added tests to cover my changes. 
- Toolbox tests already exist. Ie. QuantConnect.Tests.Toolbox.RandomValueGenerator. Additionally, I attempted to reproduce the original bug on my private fork and no bugs occurred. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: ToolBox/RandomDataGenerator/RandomDataGeneratorProgram.cs [hunk 1]
```diff
@@ -91,9 +91,12 @@ public static void GenerateRandomData(RandomDataGeneratorSettings settings, Cons
             foreach (var currentSymbol in symbolGenerator.GenerateRandomSymbols())
             {
                 // This is done so that we can update the symbol in the case of a rename event
+                TimeSpan span = settings.End.Subtract(settings.Start);
+                double diff_span = -(span.Minutes/2);
+                DateTime start_time = settings.End.AddMinutes(Math.Round(diff_span, 2, MidpointRounding.ToEven));
+                var delistDate = randomValueGenerator.NextDate(start_time, settings.End, null);
```
Comment (Martin-Molinero) [inline ToolBox/RandomDataGenerator/RandomDataGeneratorProgram.cs:97]: Cool!
Think it would be great to add some unit tests around `GenerateRandomData`, it has none 💔, at least around this particular bug, the delisting date

Comment (jovuribe) [reply ToolBox/RandomDataGenerator/RandomDataGeneratorProgram.cs]: Added the unit test :)

File: Tests/ToolBox/RandomDataGenerator/RandomDataGeneratorProgramTests.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            var delistIsLessThanEnd = true;
+            var startIsLessThanEnd = true;
+            var randomValueGenerator = new RandomValueGenerator();
+            var delistDate = RandomDataGeneratorProgram.GetDelistDate(start, end, randomValueGenerator);
+
+            //start must be less than or equal to end
+            if(start > end)
+            {
+                startIsLessThanEnd = false;
+            }
```
Comment (Martin-Molinero) [inline Tests/ToolBox/RandomDataGenerator/RandomDataGeneratorProgramTests.cs:29]: Not sure about this check, you're asserting the test arguments which should be okay already. Suggest directly adding expected delisting date and assert on it: `Assert.AreEqualt(expectedDelistingDate, delistingDate);`, will make the test more clear 👍 
nit - should remove unused imports `//using` & missing license header file

**Review by Martin-Molinero — COMMENTED** [review]
Thanks for the test, almost there

File: Tests/ToolBox/RandomDataGenerator/RandomDataGeneratorProgramTests.cs (hunk 2 above)
Comment (jovuribe) [reply Tests/ToolBox/RandomDataGenerator/RandomDataGeneratorProgramTests.cs]: I added a unit test for the midpoint value between start and end. It is difficult to directly create a TestCase for expected delist date as it is randomly generated. However, a test is done to check if the randomly generated value is between the midpoint and end.

**Review by Martin-Molinero — APPROVED** [review]
Thanks!
